feat: implement the lazy-deploy mode

doudou opened the pull request:
The current network generation + adaptation algorithm currently:
- generates a fully formed network in network generation, which means:
  * generating the task tree (compositions + task contexts)
  * replacing the task contexts by deployed tasks (via early deploy or late deploy)
- adapt the running network to match the generated network
  * which means, when running components are reused, replacing *again* tasks with the proxy of the deployed tasks from the plan

With early deploy, it means instantiating more deployments and deployed tasks than the final number of task contexts (because early deploy happens before merge), to then - in practice - replace most of them by the tasks from the plan (assuming we don't change the whole system all the time).

The late deploy mode tries to avoid all that wasted energy. Deployment in system network generation is now only setting the `orocos_name` field. We generate new deployed tasks only during the plan adaptation, and only when needed.

doudou committed:
feat: implement lazy-deploy on the system network generation

The 'lazy-deploy' mode only sets the orocos_name argument during network
generation, to avoid unnecessary creation of deployment and deployed task
instances on networks that are being modified.

This commit is the first of the overall implementation. It modifies the network
generation itself, but is not functional yet.

doudou committed:
chore: extract the adapt-runtime-plan part of the algorithm out of Engine

It will have to be significantly modified for the lazy-deploy mode, so best have
the current known-to-work version in a cleanly separated class.

doudou committed:
chore: documentation

doudou committed:
fix: use orocos_name only in MergeSolver#may_merge_task_contexts?

The lazy deployments are only settings orocos_name. The objective of the
execution_agent test is to actually check that the task's deployment has been
chosen and is the same (even if the instance is different), which can be done by
comparing orocos_name

Modify it. This fixes merging of tasks with lazy-deploy enabled

doudou committed:
fix: make sure the tasks of stubbed deployment models have names

A requirement for lazy deployment

doudou committed:
feat: modify runtime network adaptation to handle the lazy deployment as well

In the end, instead of creating a brand new copy of RuntimeNetworkAdaptation, it
seemed viable to do small adaptations to make it work in both eager and lazy
deploy cases

doudou committed:
fix: propagate the lazy_deploy flag when validating deployments from Engine and Generator

doudou committed:
fix: use orocos_name instead of execution_agent to find deployed tasks in DataflowDynamics

doudou committed:
fix: display of the TEST_LOG_LEVEL

doudou committed:
fix: make the syskit configuration precedence relation strong

The relation is here to make sure we don't execute two instances of a task
context at the same time. This *definitely* cannot be removed automatically

doudou committed:
chore: add syskit-netgen: prefix to some timepoints that were missing it

doudou committed:
fix: clear up relations only on finished tasks

The non-reusable but not finished tasks are handled normally through the
adaptation mechanisms.

doudou committed:
chore: fix complexity cops in RuntimeNetworkAdaptation

doudou commented:
@wvmcastro @jhonasiv I *strongly* recommend to first do a review commit-by-commit. I refactored Engine to extract the runtime adaptation bits, which is a good thing, but made the mistake to not spend the effort to have a clean PR with just this.

Looking commit-by-commit would allow you to see what actually is being changed on that part of the code between the original code and the one that is "ready" for lazy deployt

doudou committed:
fix: duplicate deployments in the non-lazy but early deployment case

In that case, we need to 'exfiltrate' the used_deployments hash from the network
generation step to pass to runtime network adaptation.

jhonasiv commented on lib/syskit/network_generation/system_network_deployer.rb:
```diff
@@ -250,11 +275,62 @@ def verify_all_tasks_deployed(error_handler: RaiseErrorHandler.new)
+            # @param [Component=>DeploymentGroup] deployment_groups which
+            #   deployment groups has been used for which task. This is used
+            #   to generate the error messages when needed.
+            def self.verify_all_tasks_lazily_deployed(
```
I dont see the use of this methods. the changes you made on `verify_all-tasks_deployed` is enough IMO

doudou replied:
removed

jhonasiv commented on lib/syskit/network_generation/engine.rb:
```diff
@@ -127,7 +127,7 @@ def compute_deployed_network(
                     )
 
-                    deployer.deploy(
+                    @used_deployments, = deployer.deploy(
```
Why not forward the used deployments up until the calling places of `apply_system_network_to_plan` (the `resolve` method and async AFAIK)? It feels cleaner even if I cant find another behavior problem that comes from the way you implemented it

doudou replied:
The amount of things we would need to "remove" the state from Engine itself is ... big actually.

I think we'll need sometimes to either do all of it or none of it, but right now it is 80% inside the task and 20% as argument... I don't think there's a "winner" w.r.t. being consistent.

wvmcastro commented on lib/syskit/network_generation/merge_solver.rb:
```diff
@@ -235,8 +235,9 @@ def may_merge_components?(merged_task, task)
-            # @param [Syskit::TaskContext] task
-            # @param [Syskit::TaskContext] target_task
+            # @param [Syskit::TaskContext] merged_task the task that we will merge into
+            #   the other one. This task disappears after the merge
```
"this task disapperars after the merge" is gold

wvmcastro commented on lib/syskit/network_generation/system_network_deployer.rb:
```diff
@@ -250,11 +285,62 @@ def verify_all_tasks_deployed(error_handler: RaiseErrorHandler.new)
+                query = plan.find_local_tasks(TaskContext)
+                            .not_finished.not_abstract
+                not_deployed =
+                    if lazy
+                        query.find_all { !_1.orocos_name }
+                    else
+                        query.find_all { !_1.execution_agent }
+                    end
```
Registering what we've talked in person: Create a `#deployed?` predicate to abstract what being deployed means

doudou replied:
Done.

wvmcastro reviewed (commented):
I think you could reuse the computed _used_deployments_ between `#compute_system_network` and `#compute_deployed_network` at `#resolve_system_network` when early deploy is on.

doudou committed:
fix: create well-named predicates instead of testing for orocos_name to be nil/non-nil

doudou committed:
fix: update a task's orogen_model when we select it even in lazy mode

doudou committed:
fix: instanciate scheduler tasks ("master tasks") during lazy deploy

They are actual components, not deployments, and are needed for dataflow
analysis

doudou committed:
chore: disable Style/InverseMethod for a class ordering test

I did not change the !( ... <= ... ) into ( ... > ... ) as even if Ruby somehow
pretends the order is total, it really is not. That is, I would expect A < B
and A > B to both return false (or even nil) for unrelated classes.

doudou committed:
fix: duplicate scheduler tasks appearing after lazy network adaptation of master/slave setups

doudou committed:
fix: separate initial information gathering from resolution of triggering inputs in dataflow computation

When master/slave configurations are involved, initial_information actually does
nothing for slaves (they are processed with the master, added a comment to
explain why). But then, we call triggering_inputs, which may call done_port_info
on output ports for which there are no triggering inputs.

Then when the master gets processed, the slave's initial_task_information tries
to "create" the port info on a done port, which raises.

This worked so far because, by happenstance, master tasks were always added
before slave tasks in the final deployed network. The lazy deployments changed
the order and now this fails.
